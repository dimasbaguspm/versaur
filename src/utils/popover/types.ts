import type { HTMLAttributes, ReactNode, RefObject } from 'react'

export type PopoverPlacement = 'top' | 'right' | 'bottom' | 'left'

/**
 * Popover container max-width variants
 */
export type PopoverMaxWidth = 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'full'

/**
 * Props for Popover component
 *
 * Positioned popover relative to trigger element with intelligent boundary detection.
 */
export interface PopoverProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * Unique identifier for the popover
   * This is used to link the trigger and popover via popovertarget
   */
  id: string

  /**
   * The content to display in the popover
   */
  children: ReactNode

  /**
   * Trigger element ref for positioning
   * Required for calculating popover position
   */
  triggerRef: RefObject<HTMLElement | null>

  /**
   * Whether the popover is open (controlled)
   * @default undefined (uncontrolled)
   */
  isOpen?: boolean

  /**
   * Callback when the popover opens
   */
  onOpen?: () => void

  /**
   * Callback when the popover closes
   */
  onClose?: () => void

  /**
   * Anchor element for positioning
   * Uses CSS anchor positioning when supported
   */
  anchor?: string

  /**
   * Popover behavior type
   * - auto: light dismiss (closes on outside click, Escape)
   * - manual: requires explicit close
   * @default 'auto'
   */
  popover?: 'auto' | 'manual'

  /**
   * Placement relative to trigger
   * @default 'bottom'
   */
  placement?: PopoverPlacement

  /**
   * Gap between trigger and popover in pixels
   * @default 8
   */
  gap?: number

  /**
   * Maximum width of the popover container
   * @default 'sm'
   */
  maxWidth?: PopoverMaxWidth
}

/**
 * Trigger props returned by getTriggerProps
 */
export interface PopoverTriggerProps {
  /**
   * The popover target attribute
   */
  popoverTarget: string

  /**
   * The popover target action (toggle, show, hide)
   * @default 'toggle'
   */
  popoverTargetAction?: 'toggle' | 'show' | 'hide'
}

/**
 * Return type of usePopover hook
 */
export interface UsePopoverReturn {
  /**
   * Props to spread on the trigger element
   */
  getTriggerProps: (action?: 'toggle' | 'show' | 'hide') => PopoverTriggerProps

  /**
   * Programmatically show the popover
   */
  show: () => void

  /**
   * Programmatically hide the popover
   */
  hide: () => void

  /**
   * Programmatically toggle the popover
   */
  toggle: () => void

  /**
   * Whether the popover is currently open
   */
  isOpen: boolean

  /**
   * Ref to the popover element
   */
  popoverRef: React.RefObject<HTMLDivElement | null>
}
