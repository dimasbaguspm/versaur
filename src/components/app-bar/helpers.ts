// Utility for AppBar variants (cva)
import { cva } from 'class-variance-authority'

export const appBarVariants = cva(
  'w-full flex items-center px-4 py-3 gap-2 bg-white',
  {
    variants: {
      variant: {
        primary: '',
        secondary: '',
        tertiary: '',
        ghost: '',
        neutral: '',
      },
    },
    defaultVariants: {
      variant: 'primary',
    },
  }
)

export const appBarCenterVariants = cva(
  'flex flex-col justify-center w-full min-w-0 ',
  {
    variants: {
      placement: {
        start: 'items-start',
        center: 'items-center',
        end: 'items-end',
      },
    },
    defaultVariants: {
      placement: 'start',
    },
  }
)
