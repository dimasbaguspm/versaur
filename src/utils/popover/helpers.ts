import { cva } from 'class-variance-authority'
import type { PopoverPlacement } from './types'

/**
 * Popover variants using CVA
 * Base styles for positioned popovers
 */
export const popoverVariants = cva(
  [
    // Base styles
    'fixed',
    'overflow-auto',
  ],
  {
    variants: {
      maxWidth: {
        xs: 'max-w-xs',
        sm: 'max-w-sm',
        md: 'max-w-md',
        lg: 'max-w-lg',
        xl: 'max-w-xl',
        full: 'max-w-full',
      },
    },
    defaultVariants: {
      maxWidth: 'sm',
    },
  }
)

/**
 * Calculate popover position based on trigger element and placement
 */
export function calculatePopoverPosition(
  triggerEl: HTMLElement,
  popoverEl: HTMLDivElement,
  placement: PopoverPlacement = 'bottom',
  gap: number = 8
): React.CSSProperties {
  const triggerRect = triggerEl.getBoundingClientRect()
  const popoverRect = popoverEl.getBoundingClientRect()

  let top = 0
  let left = 0

  // Calculate base position based on placement
  switch (placement) {
    case 'top':
      top = triggerRect.top - popoverRect.height - gap
      left = triggerRect.left + triggerRect.width / 2 - popoverRect.width / 2
      break
    case 'right':
      top = triggerRect.top + triggerRect.height / 2 - popoverRect.height / 2
      left = triggerRect.right + gap
      break
    case 'bottom':
      top = triggerRect.bottom + gap
      left = triggerRect.left + triggerRect.width / 2 - popoverRect.width / 2
      break
    case 'left':
      top = triggerRect.top + triggerRect.height / 2 - popoverRect.height / 2
      left = triggerRect.left - popoverRect.width - gap
      break
  }

  // Adjust to keep within viewport
  const viewportWidth = window.innerWidth
  const viewportHeight = window.innerHeight
  const margin = 8

  // Horizontal bounds
  if (left + popoverRect.width > viewportWidth - margin) {
    left = viewportWidth - popoverRect.width - margin
  }
  if (left < margin) {
    left = margin
  }

  // Vertical bounds
  if (top + popoverRect.height > viewportHeight - margin) {
    // Try flipping to opposite side
    if (placement === 'bottom') {
      top = triggerRect.top - popoverRect.height - gap
    } else if (top < margin) {
      top = margin
    }
  }
  if (top < margin) {
    // Try flipping to opposite side
    if (placement === 'top') {
      top = triggerRect.bottom + gap
    } else {
      top = margin
    }
  }

  return {
    top: `${Math.max(margin, top)}px`,
    left: `${Math.max(margin, left)}px`,
    margin: 0,
  }
}
