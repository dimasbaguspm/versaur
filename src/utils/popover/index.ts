export { Popover } from './popover'
export { usePopover } from './use-popover'
export type {
  PopoverProps,
  PopoverPlacement,
  PopoverMaxWidth,
  PopoverTriggerProps,
  UsePopoverReturn,
} from './types'
