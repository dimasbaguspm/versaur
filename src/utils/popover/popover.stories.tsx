/**
 * Popover stories for Storybook
 * Group: Utils/Popover
 */

import { useState, useRef } from 'react'
import { Popover, usePopover } from './index'

export default {
  title: 'Utils/Popover',
  parameters: {
    docs: {
      description: {
        component:
          'Browser native Popover API wrapper. Positioned relative to trigger element with intelligent viewport boundary detection.',
      },
    },
  },
}

/**
 * Uncontrolled - Basic usage with all placement options.
 * Uses browser's native popover behavior with popovertarget attribute.
 */
export const Placements = () => {
  const triggerBottomRef = useRef<HTMLButtonElement>(null)
  const triggerTopRef = useRef<HTMLButtonElement>(null)
  const triggerLeftRef = useRef<HTMLButtonElement>(null)
  const triggerRightRef = useRef<HTMLButtonElement>(null)

  return (
    <div
      style={{
        padding: '200px',
        display: 'flex',
        gap: '16px',
        flexDirection: 'column',
        alignItems: 'center',
      }}
    >
      <p style={{ marginBottom: '16px', textAlign: 'center' }}>
        Click buttons to see positioned popovers
      </p>

      <div
        style={{
          display: 'flex',
          gap: '16px',
          flexWrap: 'wrap',
          justifyContent: 'center',
        }}
      >
        <button
          ref={triggerBottomRef}
          {...Popover.getTriggerProps('popover-bottom')}
        >
          Bottom
        </button>
        <Popover
          id='popover-bottom'
          placement='bottom'
          triggerRef={triggerBottomRef}
          className='bg-primary text-background border-primary-bold'
        >
          <div style={{ padding: '12px', minWidth: '150px' }}>
            Positioned below trigger
          </div>
        </Popover>

        <button ref={triggerTopRef} {...Popover.getTriggerProps('popover-top')}>
          Top
        </button>
        <Popover
          id='popover-top'
          placement='top'
          triggerRef={triggerTopRef}
          className='bg-secondary text-background border-secondary-bold'
        >
          <div style={{ padding: '12px', minWidth: '150px' }}>
            Positioned above trigger
          </div>
        </Popover>

        <button
          ref={triggerLeftRef}
          {...Popover.getTriggerProps('popover-left')}
        >
          Left
        </button>
        <Popover
          id='popover-left'
          placement='left'
          triggerRef={triggerLeftRef}
          className='bg-tertiary text-background border-tertiary-bold'
        >
          <div style={{ padding: '12px', minWidth: '150px' }}>
            Positioned left of trigger
          </div>
        </Popover>

        <button
          ref={triggerRightRef}
          {...Popover.getTriggerProps('popover-right')}
        >
          Right
        </button>
        <Popover
          id='popover-right'
          placement='right'
          triggerRef={triggerRightRef}
          className='bg-ghost text-background border-ghost-bold'
        >
          <div style={{ padding: '12px', minWidth: '150px' }}>
            Positioned right of trigger
          </div>
        </Popover>
      </div>
    </div>
  )
}

/**
 * Controlled - Full programmatic control with useState.
 * Shows how to manage popover state externally.
 */
export const Controlled = () => {
  const triggerRef = useRef<HTMLButtonElement>(null)
  const [isOpen, setIsOpen] = useState(false)

  return (
    <div style={{ padding: '100px' }}>
      <div style={{ display: 'flex', gap: '8px', marginBottom: '16px' }}>
        <button ref={triggerRef} onClick={() => setIsOpen(!isOpen)}>
          {isOpen ? 'Close' : 'Open'}
        </button>
        <button onClick={() => setIsOpen(true)} disabled={isOpen}>
          Show
        </button>
        <button onClick={() => setIsOpen(false)} disabled={!isOpen}>
          Hide
        </button>
      </div>

      <p>Status: {isOpen ? 'Open' : 'Closed'}</p>

      <Popover
        id='controlled-popover'
        placement='bottom'
        triggerRef={triggerRef}
        isOpen={isOpen}
        onOpen={() => setIsOpen(true)}
        onClose={() => setIsOpen(false)}
        className='bg-info text-background border-info-bold'
      >
        <div style={{ padding: '16px', minWidth: '200px' }}>
          <p style={{ margin: '0 0 12px 0' }}>Controlled popover</p>
          <p style={{ margin: '0 0 12px 0', fontSize: '14px', opacity: 0.7 }}>
            State is managed externally
          </p>
          <button onClick={() => setIsOpen(false)}>Close from inside</button>
        </div>
      </Popover>
    </div>
  )
}

/**
 * With usePopover hook - Leverages the built-in hook for state management.
 * Provides programmatic controls (show, hide, toggle) and state tracking.
 */
export const WithHook = () => {
  const triggerRef = useRef<HTMLButtonElement>(null)
  const [isOpen, setIsOpen] = useState(false)

  const popover = usePopover(
    'popover-with-hook',
    isOpen,
    () => setIsOpen(true),
    () => setIsOpen(false)
  )

  return (
    <div style={{ padding: '100px' }}>
      <div style={{ display: 'flex', gap: '8px', marginBottom: '16px' }}>
        <button ref={triggerRef} {...popover.getTriggerProps()}>
          Toggle
        </button>
        <button onClick={popover.show}>Show</button>
        <button onClick={popover.hide}>Hide</button>
      </div>

      <p>Hook status: {popover.isOpen ? 'Open' : 'Closed'}</p>
      <p>State status: {isOpen ? 'Open' : 'Closed'}</p>

      <Popover
        id='popover-with-hook'
        placement='bottom'
        triggerRef={triggerRef}
        isOpen={isOpen}
        onOpen={() => setIsOpen(true)}
        onClose={() => setIsOpen(false)}
        className='bg-success text-background border-success-bold'
      >
        <div style={{ padding: '16px', minWidth: '200px' }}>
          <p style={{ margin: '0 0 12px 0' }}>usePopover hook example</p>
          <p style={{ margin: '0 0 12px 0', fontSize: '14px', opacity: 0.7 }}>
            Provides show(), hide(), toggle() methods
          </p>
          <button onClick={popover.hide}>Close</button>
        </div>
      </Popover>
    </div>
  )
}
