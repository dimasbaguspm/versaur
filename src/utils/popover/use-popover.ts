import { useRef, useEffect, useState, useCallback } from 'react'
import type { UsePopoverReturn } from './types'

/**
 * usePopover - Hook for managing popover state and interactions
 *
 * @param id - Unique identifier for the popover
 * @param isOpen - Controlled open state (optional)
 * @param onOpen - Callback when popover opens
 * @param onClose - Callback when popover closes
 * @returns Popover utilities and state
 */
export function usePopover(
  id: string,
  isOpen?: boolean,
  onOpen?: () => void,
  onClose?: () => void
): UsePopoverReturn {
  const popoverRef = useRef<HTMLDivElement>(null)
  const [internalIsOpen, setInternalIsOpen] = useState(false)

  // Use controlled state if provided, otherwise use internal state
  const isControlled = isOpen !== undefined
  const open = isControlled ? isOpen : internalIsOpen

  const show = useCallback(() => {
    if (popoverRef.current) {
      popoverRef.current.showPopover?.()
    }
  }, [])

  const hide = useCallback(() => {
    if (popoverRef.current) {
      popoverRef.current.hidePopover?.()
    }
  }, [])

  const toggle = useCallback(() => {
    if (popoverRef.current) {
      popoverRef.current.togglePopover?.()
    }
  }, [])

  const getTriggerProps = useCallback(
    (action: 'toggle' | 'show' | 'hide' = 'toggle') => ({
      popoverTarget: id,
      popoverTargetAction: action,
    }),
    [id]
  )

  // Listen to popover toggle events
  useEffect(() => {
    const popover = popoverRef.current
    if (!popover) return

    const handleToggle = (event: Event) => {
      const toggleEvent = event as ToggleEvent
      const newOpenState = toggleEvent.newState === 'open'

      if (!isControlled) {
        setInternalIsOpen(newOpenState)
      }

      if (newOpenState && onOpen) {
        onOpen()
      } else if (!newOpenState && onClose) {
        onClose()
      }
    }

    popover.addEventListener('toggle', handleToggle)
    return () => {
      popover.removeEventListener('toggle', handleToggle)
    }
  }, [isControlled, onOpen, onClose])

  // Sync controlled state with popover element
  useEffect(() => {
    if (!isControlled || !popoverRef.current) return

    const popover = popoverRef.current

    // Check if popover API is supported (may not be in test environments)
    if (!popover.showPopover || !popover.hidePopover) return

    // Try to check if popover is open using matches
    // In test environments, this might fail, so we wrap in try-catch
    let matches = false
    try {
      matches = popover.matches(':popover-open')
    } catch {
      // In environments without :popover-open support (like jsdom),
      // we can't reliably check the state, so we'll just attempt to sync
      // This is acceptable for testing purposes
    }

    if (isOpen && !matches) {
      show()
    } else if (!isOpen && matches) {
      hide()
    }
  }, [isOpen, isControlled, show, hide])

  return {
    getTriggerProps,
    show,
    hide,
    toggle,
    isOpen: open,
    popoverRef,
  }
}
