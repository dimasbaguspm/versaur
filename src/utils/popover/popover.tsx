import React, { forwardRef, useEffect, useState } from 'react'
import { cn } from '@/utils/cn'
import { popoverVariants, calculatePopoverPosition } from './helpers'
import { usePopover } from './use-popover'
import type { PopoverProps } from './types'
import { combineRefs } from '../combine-ref'

/**
 * Popover component using the browser Popover API
 *
 * Positioned relative to trigger element with intelligent viewport boundary detection.
 *
 * @example
 * ```tsx
 * // Uncontrolled usage
 * const triggerRef = useRef(null)
 * <button ref={triggerRef} {...Popover.getTriggerProps('my-popover')}>Open</button>
 * <Popover id="my-popover" triggerRef={triggerRef} placement="bottom">
 *   <p>Popover content</p>
 * </Popover>
 *
 * // Controlled usage
 * const [isOpen, setIsOpen] = useState(false)
 * const triggerRef = useRef(null)
 * <button ref={triggerRef} onClick={() => setIsOpen(true)}>Open</button>
 * <Popover
 *   id="my-popover"
 *   triggerRef={triggerRef}
 *   isOpen={isOpen}
 *   onClose={() => setIsOpen(false)}
 * >
 *   <p>Content</p>
 * </Popover>
 * ```
 */
const PopoverComponent = forwardRef<HTMLDivElement, PopoverProps>(
  (
    {
      id,
      children,
      isOpen,
      onOpen,
      onClose,
      anchor,
      popover = 'auto',
      placement = 'bottom',
      triggerRef,
      gap = 8,
      maxWidth = 'sm',
      className,
      style,
      ...props
    },
    ref
  ) => {
    const { popoverRef, isOpen: actualIsOpen } = usePopover(
      id,
      isOpen,
      onOpen,
      onClose
    )
    const [inlinePosition, setInlinePosition] = useState<React.CSSProperties>(
      {}
    )
    const [isPositioned, setIsPositioned] = useState(false)

    // Calculate inline positioning
    useEffect(() => {
      if (!triggerRef?.current || !popoverRef.current) {
        return
      }

      const updatePosition = () => {
        if (!triggerRef.current || !popoverRef.current) return

        const position = calculatePopoverPosition(
          triggerRef.current,
          popoverRef.current,
          placement,
          gap
        )
        setInlinePosition(position)
      }

      // When popover opens, wait for render, then calculate position, then show
      if (actualIsOpen) {
        setIsPositioned(false)

        // Wait for popover to be rendered in DOM
        requestAnimationFrame(() => {
          requestAnimationFrame(() => {
            // Now calculate position with actual dimensions
            updatePosition()
            // Wait one more frame before showing
            requestAnimationFrame(() => {
              setIsPositioned(true)
            })
          })
        })
      } else {
        setIsPositioned(false)
        // Calculate position while closed for initial setup
        updatePosition()
      }

      // Update on scroll and resize only when popover is open
      const handleUpdate = () => {
        if (actualIsOpen) {
          updatePosition()
        }
      }

      window.addEventListener('scroll', handleUpdate, true)
      window.addEventListener('resize', handleUpdate)

      return () => {
        window.removeEventListener('scroll', handleUpdate, true)
        window.removeEventListener('resize', handleUpdate)
      }
    }, [triggerRef, popoverRef, placement, gap, actualIsOpen])

    const combinedStyles = {
      ...inlinePosition,
      ...style,
      // Hide until positioned to prevent flash
      ...(isPositioned
        ? {}
        : {
            opacity: 0,
            pointerEvents: 'none' as const,
            transition: 'none',
          }),
    } as React.CSSProperties

    return (
      <div
        {...props}
        ref={combineRefs(ref, popoverRef)}
        id={id}
        popover={popover}
        className={cn(popoverVariants({ maxWidth }), className)}
        style={combinedStyles}
        {...(anchor && { anchor })}
      >
        {children}
      </div>
    )
  }
)

PopoverComponent.displayName = 'Popover'

/**
 * Static method to get trigger props for uncontrolled usage
 */
const getTriggerProps = (
  id: string,
  action: 'toggle' | 'show' | 'hide' = 'toggle'
) => ({
  popoverTarget: id,
  popoverTargetAction: action,
})

export const Popover = Object.assign(PopoverComponent, {
  getTriggerProps,
})
