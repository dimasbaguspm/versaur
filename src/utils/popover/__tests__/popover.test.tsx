import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest'
import { render, screen, waitFor } from '@testing-library/react'
import { Popover } from '../popover'
import { usePopover } from '../use-popover'
import { calculatePopoverPosition } from '../helpers'
import { useRef, useState } from 'react'

import * as stories from '../popover.stories'
import { composeStories } from '@storybook/react'

const { Placements, Controlled, WithHook } = composeStories(stories)

// Mock requestAnimationFrame for tests
beforeEach(() => {
  vi.spyOn(window, 'requestAnimationFrame').mockImplementation(cb => {
    cb(0)
    return 0
  })
})

afterEach(() => {
  vi.restoreAllMocks()
})

describe('Popover', () => {
  it('renders correctly with required props', () => {
    const TestComponent = () => {
      const triggerRef = useRef<HTMLButtonElement>(null)

      return (
        <>
          <button ref={triggerRef}>Trigger</button>
          <Popover id='test-popover' triggerRef={triggerRef}>
            <p>Popover content</p>
          </Popover>
        </>
      )
    }

    render(<TestComponent />)

    expect(screen.getByText('Trigger')).toBeInTheDocument()
    const popover = document.getElementById('test-popover')
    expect(popover).toBeInTheDocument()
    expect(popover).toHaveAttribute('popover', 'auto')
    expect(screen.getByText('Popover content')).toBeInTheDocument()
  })

  it('matches snapshot', () => {
    const TestComponent = () => {
      const triggerRef = useRef<HTMLButtonElement>(null)
      return (
        <>
          <button ref={triggerRef}>Trigger</button>
          <Popover id='test-popover' triggerRef={triggerRef}>
            <p>Content</p>
          </Popover>
        </>
      )
    }

    const { asFragment } = render(<TestComponent />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('applies correct popover attribute', () => {
    const TestComponent = () => {
      const triggerRef = useRef<HTMLButtonElement>(null)
      return (
        <>
          <button ref={triggerRef}>Trigger</button>
          <Popover id='manual-popover' triggerRef={triggerRef} popover='manual'>
            Content
          </Popover>
        </>
      )
    }

    render(<TestComponent />)
    const popover = document.getElementById('manual-popover')
    expect(popover).toHaveAttribute('popover', 'manual')
  })

  it('applies custom className', () => {
    const TestComponent = () => {
      const triggerRef = useRef<HTMLButtonElement>(null)
      return (
        <>
          <button ref={triggerRef}>Trigger</button>
          <Popover
            id='styled-popover'
            triggerRef={triggerRef}
            className='custom-class'
          >
            Content
          </Popover>
        </>
      )
    }
    render(<TestComponent />)
    const popover = document.getElementById('styled-popover')
    expect(popover).toHaveClass('custom-class')
  })

  it('applies inline styles', () => {
    const TestComponent = () => {
      const triggerRef = useRef<HTMLButtonElement>(null)
      return (
        <>
          <button ref={triggerRef}>Trigger</button>
          <Popover
            id='styled-popover'
            triggerRef={triggerRef}
            style={{ zIndex: 9999 }}
          >
            Content
          </Popover>
        </>
      )
    }

    render(<TestComponent />)
    const popover = document.getElementById('styled-popover')
    expect(popover).toHaveStyle({ zIndex: '9999' })
  })

  it('forwards ref correctly', () => {
    const TestComponent = () => {
      const triggerRef = useRef<HTMLButtonElement>(null)
      const popoverRef = useRef<HTMLDivElement>(null)

      return (
        <>
          <button ref={triggerRef}>Trigger</button>
          <Popover id='test-popover' triggerRef={triggerRef} ref={popoverRef}>
            Content
          </Popover>
        </>
      )
    }

    render(<TestComponent />)
    const popover = document.getElementById('test-popover')
    expect(popover).toBeInTheDocument()
  })

  it('generates trigger props with getTriggerProps', () => {
    const props = Popover.getTriggerProps('my-popover')
    expect(props).toEqual({
      popoverTarget: 'my-popover',
      popoverTargetAction: 'toggle',
    })
  })

  it('generates trigger props with custom action', () => {
    const showProps = Popover.getTriggerProps('my-popover', 'show')
    expect(showProps).toEqual({
      popoverTarget: 'my-popover',
      popoverTargetAction: 'show',
    })

    const hideProps = Popover.getTriggerProps('my-popover', 'hide')
    expect(hideProps).toEqual({
      popoverTarget: 'my-popover',
      popoverTargetAction: 'hide',
    })
  })

  it('handles controlled state', () => {
    const TestComponent = () => {
      const triggerRef = useRef<HTMLButtonElement>(null)
      const [isOpen, setIsOpen] = useState(false)

      return (
        <>
          <button ref={triggerRef} onClick={() => setIsOpen(true)}>
            Open
          </button>
          <Popover
            id='controlled-popover'
            triggerRef={triggerRef}
            isOpen={isOpen}
            onClose={() => setIsOpen(false)}
          >
            Content
          </Popover>
        </>
      )
    }

    render(<TestComponent />)
    const popover = document.getElementById('controlled-popover')
    expect(popover).toBeInTheDocument()
  })

  it('accepts anchor attribute', () => {
    const TestComponent = () => {
      const triggerRef = useRef<HTMLButtonElement>(null)
      return (
        <>
          <button ref={triggerRef}>Trigger</button>
          <Popover
            id='anchored-popover'
            triggerRef={triggerRef}
            anchor='--my-anchor'
          >
            Content
          </Popover>
        </>
      )
    }

    render(<TestComponent />)
    const popover = document.getElementById('anchored-popover')
    expect(popover).toHaveAttribute('anchor', '--my-anchor')
  })

  it('calculates position for all placements', async () => {
    const placements = ['top', 'right', 'bottom', 'left'] as const

    for (const placement of placements) {
      const TestComponent = () => {
        const triggerRef = useRef<HTMLButtonElement>(null)
        return (
          <>
            <button ref={triggerRef}>Trigger</button>
            <Popover
              id={`popover-${placement}`}
              triggerRef={triggerRef}
              placement={placement}
            >
              Content
            </Popover>
          </>
        )
      }

      const { unmount } = render(<TestComponent />)
      const popover = document.getElementById(`popover-${placement}`)
      expect(popover).toBeInTheDocument()

      // Position should be set in style attribute
      await waitFor(() => {
        expect(popover).toHaveAttribute('style')
        const style = popover?.getAttribute('style') || ''
        expect(style).toContain('top:')
        expect(style).toContain('left:')
      })

      unmount()
    }
  })

  it('renders Placements story correctly', () => {
    render(<Placements />)

    // Check all placement buttons are rendered
    expect(screen.getByText('Top')).toBeInTheDocument()
    expect(screen.getByText('Right')).toBeInTheDocument()
    expect(screen.getByText('Bottom')).toBeInTheDocument()
    expect(screen.getByText('Left')).toBeInTheDocument()

    // Check all popover contents
    expect(screen.getByText('Positioned above trigger')).toBeInTheDocument()
    expect(screen.getByText('Positioned right of trigger')).toBeInTheDocument()
    expect(screen.getByText('Positioned below trigger')).toBeInTheDocument()
    expect(screen.getByText('Positioned left of trigger')).toBeInTheDocument()
  })

  it('renders Controlled story correctly', () => {
    render(<Controlled />)
    expect(screen.getByText('Open')).toBeInTheDocument()
    expect(screen.getByText('Show')).toBeInTheDocument()
    expect(screen.getByText('Hide')).toBeInTheDocument()
    expect(screen.getByText('Controlled popover')).toBeInTheDocument()
  })

  it('renders WithHook story correctly', () => {
    render(<WithHook />)
    expect(screen.getByText('Toggle')).toBeInTheDocument()
    expect(screen.getByText('Show')).toBeInTheDocument()
    expect(screen.getByText('Hide')).toBeInTheDocument()
    expect(screen.getByText('usePopover hook example')).toBeInTheDocument()
  })
})

describe('usePopover', () => {
  it('returns correct interface', () => {
    const TestComponent = () => {
      const popover = usePopover('test-id')

      expect(popover).toHaveProperty('popoverRef')
      expect(popover).toHaveProperty('isOpen')
      expect(popover).toHaveProperty('show')
      expect(popover).toHaveProperty('hide')
      expect(popover).toHaveProperty('toggle')
      expect(popover).toHaveProperty('getTriggerProps')

      expect(typeof popover.show).toBe('function')
      expect(typeof popover.hide).toBe('function')
      expect(typeof popover.toggle).toBe('function')
      expect(typeof popover.getTriggerProps).toBe('function')

      return <div>Test</div>
    }

    render(<TestComponent />)
  })

  it('generates trigger props', () => {
    const TestComponent = () => {
      const popover = usePopover('test-id')
      const props = popover.getTriggerProps()

      expect(props).toEqual({
        popoverTarget: 'test-id',
        popoverTargetAction: 'toggle',
      })

      return <div>Test</div>
    }

    render(<TestComponent />)
  })

  it('generates trigger props with custom action', () => {
    const TestComponent = () => {
      const popover = usePopover('test-id')

      const showProps = popover.getTriggerProps('show')
      expect(showProps).toEqual({
        popoverTarget: 'test-id',
        popoverTargetAction: 'show',
      })

      const hideProps = popover.getTriggerProps('hide')
      expect(hideProps).toEqual({
        popoverTarget: 'test-id',
        popoverTargetAction: 'hide',
      })

      return <div>Test</div>
    }

    render(<TestComponent />)
  })

  it('handles controlled state', () => {
    const TestComponent = () => {
      const [isOpen, setIsOpen] = useState(false)
      const popover = usePopover(
        'test-id',
        isOpen,
        () => setIsOpen(true),
        () => setIsOpen(false)
      )

      expect(popover.isOpen).toBe(false)

      return <div>Test</div>
    }

    render(<TestComponent />)
  })
})

describe('calculatePopoverPosition', () => {
  const mockTrigger = {
    getBoundingClientRect: () => ({
      top: 100,
      left: 100,
      bottom: 150,
      right: 200,
      width: 100,
      height: 50,
      x: 100,
      y: 100,
      toJSON: () => {},
    }),
  } as HTMLElement

  const mockPopover = {
    getBoundingClientRect: () => ({
      top: 0,
      left: 0,
      bottom: 100,
      right: 200,
      width: 200,
      height: 100,
      x: 0,
      y: 0,
      toJSON: () => {},
    }),
  } as HTMLDivElement

  beforeEach(() => {
    Object.defineProperty(window, 'innerWidth', {
      writable: true,
      configurable: true,
      value: 1024,
    })
    Object.defineProperty(window, 'innerHeight', {
      writable: true,
      configurable: true,
      value: 768,
    })
  })

  it('calculates position for bottom placement', () => {
    const position = calculatePopoverPosition(
      mockTrigger,
      mockPopover,
      'bottom',
      8
    )

    expect(position).toHaveProperty('top')
    expect(position).toHaveProperty('left')
    expect(position.margin).toBe(0)

    // Bottom: top = triggerBottom + gap = 150 + 8 = 158
    expect(position.top).toBe('158px')
    // Centered: left = triggerLeft + (triggerWidth / 2) - (popoverWidth / 2)
    // = 100 + 50 - 100 = 50
    expect(position.left).toBe('50px')
  })

  it('calculates position for top placement', () => {
    const position = calculatePopoverPosition(
      mockTrigger,
      mockPopover,
      'top',
      8
    )

    // Top: top = triggerTop - popoverHeight - gap = 100 - 100 - 8 = -8
    // Should be adjusted to viewport edge
    expect(position).toHaveProperty('top')
    expect(position).toHaveProperty('left')
  })

  it('calculates position for right placement', () => {
    const position = calculatePopoverPosition(
      mockTrigger,
      mockPopover,
      'right',
      8
    )

    // Right: left = triggerRight + gap = 200 + 8 = 208
    expect(position.left).toBe('208px')
    // Centered vertically
    expect(position).toHaveProperty('top')
  })

  it('calculates position for left placement', () => {
    const position = calculatePopoverPosition(
      mockTrigger,
      mockPopover,
      'left',
      8
    )

    // Left: left = triggerLeft - popoverWidth - gap = 100 - 200 - 8 = -108
    // Should be adjusted to viewport edge
    expect(position).toHaveProperty('top')
    expect(position).toHaveProperty('left')
  })

  it('handles viewport boundary detection', () => {
    // Create trigger near right edge
    const edgeTrigger = {
      getBoundingClientRect: () => ({
        top: 100,
        left: 900,
        bottom: 150,
        right: 1000,
        width: 100,
        height: 50,
        x: 900,
        y: 100,
        toJSON: () => {},
      }),
    } as HTMLElement

    const position = calculatePopoverPosition(
      edgeTrigger,
      mockPopover,
      'bottom',
      8
    )

    // Position should be adjusted to not overflow viewport
    expect(position).toHaveProperty('left')
    expect(position).toHaveProperty('top')
    expect(position.margin).toBe(0)
  })

  it('respects custom gap', () => {
    const position = calculatePopoverPosition(
      mockTrigger,
      mockPopover,
      'bottom',
      16
    )

    // Bottom with gap 16: top = 150 + 16 = 166
    expect(position.top).toBe('166px')
  })
})
