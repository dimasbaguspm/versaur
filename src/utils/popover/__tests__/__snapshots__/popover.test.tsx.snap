// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Popover > matches snapshot 1`] = `
<DocumentFragment>
  <button>
    Trigger
  </button>
  <div
    class="fixed overflow-auto max-w-sm"
    id="test-popover"
    popover="auto"
    style="opacity: 0; pointer-events: none; transition: none; top: 8px; left: 8px; margin: 0px;"
  >
    <p>
      Content
    </p>
  </div>
</DocumentFragment>
`;
